// snstr/dist/src/nip46/index.js
"use strict";
/**
 * NIP-46 Remote Signing Protocol
 *
 * This implementation provides a way to manage private keys remotely
 * and sign events from different clients. The protocol establishes a secure
 * connection between clients and a "bunker" which holds the private keys.
 *
 * Reference: https://github.com/nostr-protocol/nips/blob/master/46.md
 *
 * Two implementations are provided:
 *
 * 1. NostrRemoteSignerClient/NostrRemoteSignerBunker:
 *    - Full-featured implementation with all NIP-46 capabilities
 *    - Supports auth challenges via URL
 *    - Handles permissions for specific event kinds
 *    - Supports both NIP-04 and NIP-44 encryption
 *    - Connection metadata for better UX
 *    - Secret token support for secure connections
 *
 * 2. SimpleNIP46Client/SimpleNIP46Bunker:
 *    - Simplified implementation for basic use cases
 *    - Minimal code with core functionality
 *    - NIP-04 encryption only
 *    - Basic permission handling
 *    - No auth challenges or metadata support
 *    - Good for learning and simple applications
 *
 * Choose the implementation that best fits your needs based on the complexity
 * and security requirements of your application.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateKeypair = exports.SecureErrorHandler = exports.validateTimestamp = exports.sanitizeString = exports.validateAndParseJson = exports.validateConnectionString = exports.validatePermission = exports.validateRelayUrl = exports.validateParams = exports.isValidMethod = exports.validateRequestPayload = exports.validatePrivateKey = exports.validateSignature = exports.validateEventId = exports.validatePubkey = exports.validateEventContent = exports.MAX_TAG_ELEMENT_LENGTH = exports.MAX_TAGS_COUNT = exports.MAX_PARAM_LENGTH = exports.MAX_PARAMS_COUNT = exports.MAX_ID_LENGTH = exports.MAX_CONTENT_SIZE = exports.parseConnectionString = exports.buildConnectionString = exports.createErrorResponse = exports.createSuccessResponse = exports.createRequest = exports.initializeCrypto = exports.generateRequestId = exports.LogLevel = exports.Logger = exports.SimpleNIP46Bunker = exports.SimpleNIP46Client = exports.NostrRemoteSignerBunker = exports.NostrRemoteSignerClient = void 0;
// Export types
__exportStar(require("./types"), exports);
// Export client and bunker implementations
var client_1 = require("./client");
Object.defineProperty(exports, "NostrRemoteSignerClient", { enumerable: true, get: function () { return client_1.NostrRemoteSignerClient; } });
var bunker_1 = require("./bunker");
Object.defineProperty(exports, "NostrRemoteSignerBunker", { enumerable: true, get: function () { return bunker_1.NostrRemoteSignerBunker; } });
// Export simplified implementation
var simple_client_1 = require("./simple-client");
Object.defineProperty(exports, "SimpleNIP46Client", { enumerable: true, get: function () { return simple_client_1.SimpleNIP46Client; } });
var simple_bunker_1 = require("./simple-bunker");
Object.defineProperty(exports, "SimpleNIP46Bunker", { enumerable: true, get: function () { return simple_bunker_1.SimpleNIP46Bunker; } });
// Export utilities
var logger_1 = require("./utils/logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return logger_1.Logger; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return logger_1.LogLevel; } });
var request_response_1 = require("./utils/request-response");
Object.defineProperty(exports, "generateRequestId", { enumerable: true, get: function () { return request_response_1.generateRequestId; } });
Object.defineProperty(exports, "initializeCrypto", { enumerable: true, get: function () { return request_response_1.initializeCrypto; } });
Object.defineProperty(exports, "createRequest", { enumerable: true, get: function () { return request_response_1.createRequest; } });
Object.defineProperty(exports, "createSuccessResponse", { enumerable: true, get: function () { return request_response_1.createSuccessResponse; } });
Object.defineProperty(exports, "createErrorResponse", { enumerable: true, get: function () { return request_response_1.createErrorResponse; } });
var connection_1 = require("./utils/connection");
Object.defineProperty(exports, "buildConnectionString", { enumerable: true, get: function () { return connection_1.buildConnectionString; } });
Object.defineProperty(exports, "parseConnectionString", { enumerable: true, get: function () { return connection_1.parseConnectionString; } });
var validator_1 = require("./utils/validator");
// Constants
Object.defineProperty(exports, "MAX_CONTENT_SIZE", { enumerable: true, get: function () { return validator_1.MAX_CONTENT_SIZE; } });
Object.defineProperty(exports, "MAX_ID_LENGTH", { enumerable: true, get: function () { return validator_1.MAX_ID_LENGTH; } });
Object.defineProperty(exports, "MAX_PARAMS_COUNT", { enumerable: true, get: function () { return validator_1.MAX_PARAMS_COUNT; } });
Object.defineProperty(exports, "MAX_PARAM_LENGTH", { enumerable: true, get: function () { return validator_1.MAX_PARAM_LENGTH; } });
Object.defineProperty(exports, "MAX_TAGS_COUNT", { enumerable: true, get: function () { return validator_1.MAX_TAGS_COUNT; } });
Object.defineProperty(exports, "MAX_TAG_ELEMENT_LENGTH", { enumerable: true, get: function () { return validator_1.MAX_TAG_ELEMENT_LENGTH; } });
// Validation functions
Object.defineProperty(exports, "validateEventContent", { enumerable: true, get: function () { return validator_1.validateEventContent; } });
Object.defineProperty(exports, "validatePubkey", { enumerable: true, get: function () { return validator_1.validatePubkey; } });
Object.defineProperty(exports, "validateEventId", { enumerable: true, get: function () { return validator_1.validateEventId; } });
Object.defineProperty(exports, "validateSignature", { enumerable: true, get: function () { return validator_1.validateSignature; } });
Object.defineProperty(exports, "validatePrivateKey", { enumerable: true, get: function () { return validator_1.validatePrivateKey; } });
Object.defineProperty(exports, "validateRequestPayload", { enumerable: true, get: function () { return validator_1.validateRequestPayload; } });
Object.defineProperty(exports, "isValidMethod", { enumerable: true, get: function () { return validator_1.isValidMethod; } });
Object.defineProperty(exports, "validateParams", { enumerable: true, get: function () { return validator_1.validateParams; } });
Object.defineProperty(exports, "validateRelayUrl", { enumerable: true, get: function () { return validator_1.validateRelayUrl; } });
Object.defineProperty(exports, "validatePermission", { enumerable: true, get: function () { return validator_1.validatePermission; } });
Object.defineProperty(exports, "validateConnectionString", { enumerable: true, get: function () { return validator_1.validateConnectionString; } });
Object.defineProperty(exports, "validateAndParseJson", { enumerable: true, get: function () { return validator_1.validateAndParseJson; } });
Object.defineProperty(exports, "sanitizeString", { enumerable: true, get: function () { return validator_1.sanitizeString; } });
Object.defineProperty(exports, "validateTimestamp", { enumerable: true, get: function () { return validator_1.validateTimestamp; } });
// Error handler class
Object.defineProperty(exports, "SecureErrorHandler", { enumerable: true, get: function () { return validator_1.SecureErrorHandler; } });
// Re-export utility functions from crypto to make the API more convenient
var crypto_1 = require("../utils/crypto");
Object.defineProperty(exports, "generateKeypair", { enumerable: true, get: function () { return crypto_1.generateKeypair; } });
