// snstr/dist/src/nip46/types.js
"use strict";
/**
 * NIP-46 request message format
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NIP46ErrorUtils = exports.NIP46ReplayAttackError = exports.NIP46SecurityError = exports.NIP46SigningError = exports.NIP46DecryptionError = exports.NIP46EncryptionError = exports.NIP46AuthorizationError = exports.NIP46TimeoutError = exports.NIP46ConnectionError = exports.NIP46Error = exports.NIP46Method = exports.NIP46ErrorCode = void 0;
/**
 * Standardized error codes for NIP-46 operations
 */
var NIP46ErrorCode;
(function (NIP46ErrorCode) {
    // Connection errors
    NIP46ErrorCode["INVALID_REQUEST"] = "INVALID_REQUEST";
    NIP46ErrorCode["CONNECTION_REJECTED"] = "CONNECTION_REJECTED";
    NIP46ErrorCode["INVALID_SECRET"] = "INVALID_SECRET";
    NIP46ErrorCode["PERMISSION_DENIED"] = "PERMISSION_DENIED";
    NIP46ErrorCode["NOT_AUTHORIZED"] = "NOT_AUTHORIZED";
    // Method errors
    NIP46ErrorCode["METHOD_NOT_SUPPORTED"] = "METHOD_NOT_SUPPORTED";
    NIP46ErrorCode["INVALID_PARAMETERS"] = "INVALID_PARAMETERS";
    // Signing errors
    NIP46ErrorCode["SIGNING_FAILED"] = "SIGNING_FAILED";
    NIP46ErrorCode["USER_PRIVATE_KEY_NOT_SET"] = "USER_PRIVATE_KEY_NOT_SET";
    // Encryption errors
    NIP46ErrorCode["ENCRYPTION_FAILED"] = "ENCRYPTION_FAILED";
    NIP46ErrorCode["DECRYPTION_FAILED"] = "DECRYPTION_FAILED";
    // General errors
    NIP46ErrorCode["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    NIP46ErrorCode["TIMEOUT"] = "TIMEOUT";
    NIP46ErrorCode["RATE_LIMITED"] = "RATE_LIMITED";
})(NIP46ErrorCode || (exports.NIP46ErrorCode = NIP46ErrorCode = {}));
/**
 * Supported NIP-46 methods
 */
var NIP46Method;
(function (NIP46Method) {
    NIP46Method["CONNECT"] = "connect";
    NIP46Method["GET_PUBLIC_KEY"] = "get_public_key";
    NIP46Method["SIGN_EVENT"] = "sign_event";
    NIP46Method["GET_RELAYS"] = "get_relays";
    NIP46Method["PING"] = "ping";
    NIP46Method["DISCONNECT"] = "disconnect";
    NIP46Method["NIP04_ENCRYPT"] = "nip04_encrypt";
    NIP46Method["NIP04_DECRYPT"] = "nip04_decrypt";
    NIP46Method["NIP44_ENCRYPT"] = "nip44_encrypt";
    NIP46Method["NIP44_DECRYPT"] = "nip44_decrypt";
})(NIP46Method || (exports.NIP46Method = NIP46Method = {}));
/**
 * Error types for NIP-46 operations
 */
class NIP46Error extends Error {
    constructor(message) {
        super(message);
        this.name = "NIP46Error";
    }
}
exports.NIP46Error = NIP46Error;
class NIP46ConnectionError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46ConnectionError";
    }
}
exports.NIP46ConnectionError = NIP46ConnectionError;
class NIP46TimeoutError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46TimeoutError";
    }
}
exports.NIP46TimeoutError = NIP46TimeoutError;
class NIP46AuthorizationError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46AuthorizationError";
    }
}
exports.NIP46AuthorizationError = NIP46AuthorizationError;
class NIP46EncryptionError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46EncryptionError";
    }
}
exports.NIP46EncryptionError = NIP46EncryptionError;
class NIP46DecryptionError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46DecryptionError";
    }
}
exports.NIP46DecryptionError = NIP46DecryptionError;
class NIP46SigningError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46SigningError";
    }
}
exports.NIP46SigningError = NIP46SigningError;
class NIP46SecurityError extends NIP46Error {
    constructor(message) {
        super(message);
        this.name = "NIP46SecurityError";
    }
}
exports.NIP46SecurityError = NIP46SecurityError;
class NIP46ReplayAttackError extends NIP46SecurityError {
    constructor(message) {
        super(message);
        this.name = "NIP46ReplayAttackError";
    }
}
exports.NIP46ReplayAttackError = NIP46ReplayAttackError;
/**
 * Utility functions for standardized error handling
 */
class NIP46ErrorUtils {
    /**
     * Create a standardized error response
     */
    static createErrorResponse(id, code, message, details) {
        const errorObj = { code, message, details };
        return {
            id,
            error: JSON.stringify(errorObj),
        };
    }
    /**
     * Create a simple error response (for backwards compatibility)
     */
    static createSimpleErrorResponse(id, message) {
        return {
            id,
            error: message,
        };
    }
    /**
     * Map error codes to HTTP-like status descriptions
     */
    static getErrorDescription(code) {
        const descriptions = {
            [NIP46ErrorCode.INVALID_REQUEST]: "The request format is invalid",
            [NIP46ErrorCode.CONNECTION_REJECTED]: "Connection was rejected by the signer",
            [NIP46ErrorCode.INVALID_SECRET]: "The provided secret is invalid",
            [NIP46ErrorCode.PERMISSION_DENIED]: "Insufficient permissions for this operation",
            [NIP46ErrorCode.NOT_AUTHORIZED]: "Client is not authorized",
            [NIP46ErrorCode.METHOD_NOT_SUPPORTED]: "The requested method is not supported",
            [NIP46ErrorCode.INVALID_PARAMETERS]: "The provided parameters are invalid",
            [NIP46ErrorCode.SIGNING_FAILED]: "Event signing failed",
            [NIP46ErrorCode.USER_PRIVATE_KEY_NOT_SET]: "User private key is not configured",
            [NIP46ErrorCode.ENCRYPTION_FAILED]: "Encryption operation failed",
            [NIP46ErrorCode.DECRYPTION_FAILED]: "Decryption operation failed",
            [NIP46ErrorCode.INTERNAL_ERROR]: "An internal error occurred",
            [NIP46ErrorCode.TIMEOUT]: "Operation timed out",
            [NIP46ErrorCode.RATE_LIMITED]: "Rate limit exceeded",
        };
        return descriptions[code] || "Unknown error";
    }
}
exports.NIP46ErrorUtils = NIP46ErrorUtils;
