// snstr/dist/esm/src/entries/index.web.js
// Web / React Native entry: excludes Node-only NIP-04 and re-exports a stub.
// This prevents Metro/Expo from pulling Node's `crypto` module.
// Export client classes
export { Nostr } from "../nip01/nostr.js";
export { Relay } from "../nip01/relay.js";
export { useWebSocketImplementation, resetWebSocketImplementation, } from "../utils/websocket.js";
export { RelayPool } from "../nip01/relayPool.js";
// Export types
export * from "../types/nostr.js";
// Export utilities
export { generateKeypair, getPublicKey, signEvent, verifySignature, } from "../utils/crypto.js";
export { getUnixTime } from "../utils/time.js";
export { createEvent, createAddressableEvent, getEventHash, } from "../nip01/event.js";
// Export NIP-02 utilities
export { createContactListEvent, parseContactsFromEvent, parseContactsFromEventWithWarnings, } from "../nip02/index.js";
// Export NIP-04 web/RN implementation (crypto-js AES-CBC)
export { encrypt as encryptNIP04, decrypt as decryptNIP04, getSharedSecret as getNIP04SharedSecret, } from "../nip04/web.js";
// Export NIP-11 utilities
export { fetchRelayInformation, supportsNIP11, relaySupportsNIPs, getRelayPaymentInfo, relayRequiresPayment, } from "../nip11/index.js";
// Export NIP-44 utilities
export { encrypt as encryptNIP44, decrypt as decryptNIP44, generateNonce as generateNIP44Nonce, getSharedSecret as getNIP44SharedSecret, constantTimeEqual as constantTimeEqual, secureWipe as secureWipeNIP44, } from "../nip44/index.js";
// Export NIP-17 utilities
export { createDirectMessage, decryptDirectMessage, DM_KIND, FILE_KIND, GIFT_WRAP_KIND, SEAL_KIND, initializeCrypto as initializeNIP17Crypto, } from "../nip17/index.js";
// Export NIP-19 utilities
export { 
// Core encoding/decoding functions
encodeBech32, decodeBech32, decode, 
// Public key (npub)
encodePublicKey, decodePublicKey, 
// Private key (nsec)
encodePrivateKey, decodePrivateKey, 
// Note ID (note)
encodeNoteId, decodeNoteId, 
// Profile (nprofile)
encodeProfile, decodeProfile, 
// types re-exported separately below
// Event (nevent)
encodeEvent, decodeEvent, 
// types re-exported separately below
// Address (naddr)
encodeAddress, decodeAddress, 
// types re-exported separately below
// Other helpers
// Enums
Prefix, TLVType, } from "../nip19/index.js";
// Export NIP-21 utilities
export { encodeNostrURI, decodeNostrURI, NOSTR_URI_PREFIX } from "../nip21/index.js";
// Export NIP-05 utilities
export { verifyNIP05, lookupNIP05, getNIP05PubKey, getNIP05Relays, } from "../nip05/index.js";
// Export NIP-07 utilities
export { hasNip07Support, getPublicKey as getNip07PublicKey, signEvent as signEventWithNip07, encryptNip04 as encryptNip04WithExtension, decryptNip04 as decryptNip04WithExtension, encryptNip44 as encryptNip44WithExtension, decryptNip44 as decryptNip44WithExtension, } from "../nip07/index.js";
// Export NIP-09 utilities
export { createDeletionRequest, parseDeletionTargets, isDeletionRequestForEvent, } from "../nip09/index.js";
// Export NIP-10 utilities
export { createReplyTags, createQuoteTag, parseThreadReferences, } from "../nip10/index.js";
// Export NIP-07 adapter
export { Nip07Nostr } from "../nip07/adapter.js";
// NIP-46 utilities are Node-oriented; omit from RN/web entry to reduce bundle size
// Consumers can import platform-appropriate implementations directly if needed.
// NIP-57: Lightning Zaps
export { createZapRequest, createZapReceipt, validateZapReceipt, parseZapSplit, calculateZapSplitAmounts, ZAP_REQUEST_KIND, ZAP_RECEIPT_KIND, } from "../nip57/index.js";
export { NostrZapClient, ZapClient, } from "../nip57/client.js";
export { fetchLnurlPayMetadata, supportsNostrZaps, decodeLnurl, buildZapCallbackUrl, extractLnurlMetadata, parseBolt11Invoice, } from "../nip57/utils.js";
// NIP-47 is Node-leaning and may pull NIP-04; omit from RN/web entry.
// NIP-50 search utilities
export { createSearchFilter } from "../nip50/index.js";
