// snstr/dist/src/nip09/index.js
"use strict";
/**
 * NIP-09: Event Deletion Request
 *
 * Utilities for creating and processing deletion request events.
 * https://github.com/nostr-protocol/nips/blob/master/09.md
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDeletionRequest = createDeletionRequest;
exports.parseDeletionTargets = parseDeletionTargets;
exports.isDeletionRequestForEvent = isDeletionRequestForEvent;
const event_1 = require("../nip01/event");
const nostr_1 = require("../types/nostr");
const security_validator_1 = require("../utils/security-validator");
/**
 * Create an unsigned deletion request event (kind 5)
 */
function createDeletionRequest(opts, pubkey) {
    const tags = [];
    for (const id of opts.ids || []) {
        tags.push(["e", id]);
    }
    for (const addr of opts.addresses || []) {
        tags.push(["a", addr]);
    }
    for (const kind of opts.kinds || []) {
        tags.push(["k", kind.toString()]);
    }
    return (0, event_1.createEvent)({
        kind: nostr_1.NostrKind.Deletion,
        content: opts.content || "",
        tags,
    }, pubkey);
}
/**
 * Extract referenced ids, addresses and kinds from a deletion event
 */
function parseDeletionTargets(event) {
    const result = {
        ids: [],
        addresses: [],
        kinds: [],
    };
    for (const tag of event.tags) {
        try {
            // Safe access to tag elements with bounds checking
            if (!(0, security_validator_1.validateArrayAccess)(tag, 0) || !(0, security_validator_1.validateArrayAccess)(tag, 1)) {
                continue; // Skip malformed tags
            }
            const tagName = (0, security_validator_1.safeArrayAccess)(tag, 0);
            const tagValue = (0, security_validator_1.safeArrayAccess)(tag, 1);
            if (typeof tagName !== "string" || typeof tagValue !== "string") {
                continue; // Skip malformed tags
            }
            if (tagName === "e" && tagValue) {
                result.ids.push(tagValue);
            }
            else if (tagName === "a" && tagValue) {
                result.addresses.push(tagValue);
            }
            else if (tagName === "k" && tagValue) {
                const k = parseInt(tagValue, 10);
                if (!isNaN(k)) {
                    result.kinds.push(k);
                }
            }
        }
        catch (error) {
            if (error instanceof security_validator_1.SecurityValidationError) {
                // Log bounds checking error but continue processing
                if (typeof console !== "undefined" && console.warn) {
                    console.warn(`NIP-09: Bounds checking error in tag processing: ${error.message}`);
                }
            }
        }
    }
    return result;
}
/**
 * Check if a deletion request targets the given event
 */
function isDeletionRequestForEvent(deletion, event) {
    if (deletion.kind !== nostr_1.NostrKind.Deletion)
        return false;
    if (deletion.pubkey !== event.pubkey)
        return false;
    const targets = parseDeletionTargets(deletion);
    if (targets.ids.includes(event.id))
        return true;
    // Safe access to d-tag with bounds checking
    try {
        const dTag = event.tags.find((t) => {
            try {
                return (0, security_validator_1.validateArrayAccess)(t, 0) && (0, security_validator_1.safeArrayAccess)(t, 0) === "d";
            }
            catch {
                return false;
            }
        });
        if (dTag && (0, security_validator_1.validateArrayAccess)(dTag, 1)) {
            const dValue = (0, security_validator_1.safeArrayAccess)(dTag, 1);
            if (typeof dValue === "string") {
                const addr = `${event.kind}:${event.pubkey}:${dValue}`;
                if (targets.addresses.includes(addr))
                    return true;
            }
        }
    }
    catch (error) {
        if (error instanceof security_validator_1.SecurityValidationError) {
            if (typeof console !== "undefined" && console.warn) {
                console.warn(`NIP-09: Bounds checking error in d-tag processing: ${error.message}`);
            }
        }
    }
    return false;
}
