// snstr/dist/esm/src/types/nostr.js
/**
 * Relay event types as specified in NIP-01
 */
export var RelayEvent;
(function (RelayEvent) {
    /** Fired when a connection to a relay is established */
    RelayEvent["Connect"] = "connect";
    /** Fired when a connection to a relay is closed */
    RelayEvent["Disconnect"] = "disconnect";
    /** Fired when an error occurs */
    RelayEvent["Error"] = "error";
    /** Fired when a relay sends a NOTICE message */
    RelayEvent["Notice"] = "notice";
    /** Fired when a relay confirms an event was received (OK message) */
    RelayEvent["OK"] = "ok";
    /** Fired when a relay closes a subscription */
    RelayEvent["Closed"] = "closed";
    /** Fired when a relay sends an authentication challenge (NIP-42) */
    RelayEvent["Auth"] = "auth";
})(RelayEvent || (RelayEvent = {}));
/**
 * Relay status indicating the current connection state
 */
export var RelayStatus;
(function (RelayStatus) {
    /** Not connected or disconnected */
    RelayStatus["DISCONNECTED"] = "disconnected";
    /** Connection in progress */
    RelayStatus["CONNECTING"] = "connecting";
    /** Successfully connected */
    RelayStatus["CONNECTED"] = "connected";
    /** Waiting to reconnect after a disconnection */
    RelayStatus["RECONNECTING"] = "reconnecting";
    /** Permanently closed (e.g., after calling disconnect) */
    RelayStatus["CLOSED"] = "closed";
})(RelayStatus || (RelayStatus = {}));
/**
 * NIP-01 defined event kinds with their semantic meanings
 * https://github.com/nostr-protocol/nips/blob/master/01.md
 */
export var NostrKind;
(function (NostrKind) {
    /** Metadata: Profile information (replaceable) */
    NostrKind[NostrKind["Metadata"] = 0] = "Metadata";
    /** Short note: Standard text note */
    NostrKind[NostrKind["ShortNote"] = 1] = "ShortNote";
    /** Recommend relay: Relay recommendations (replaceable) */
    NostrKind[NostrKind["RecommendRelay"] = 2] = "RecommendRelay";
    /** Contacts: Follow list (replaceable) */
    NostrKind[NostrKind["Contacts"] = 3] = "Contacts";
    /** Direct message: Encrypted direct message */
    NostrKind[NostrKind["DirectMessage"] = 4] = "DirectMessage";
    /** Deletion: Event deletion request */
    NostrKind[NostrKind["Deletion"] = 5] = "Deletion";
    /** Repost: Repost of another note */
    NostrKind[NostrKind["Repost"] = 6] = "Repost";
    /** Reaction: Reaction to another note */
    NostrKind[NostrKind["Reaction"] = 7] = "Reaction";
    /** Badge award: Award badge to a user */
    NostrKind[NostrKind["BadgeAward"] = 8] = "BadgeAward";
})(NostrKind || (NostrKind = {}));
/**
 * WebSocket ready states
 */
export var WebSocketReadyState;
(function (WebSocketReadyState) {
    /** Connection has not yet been established */
    WebSocketReadyState[WebSocketReadyState["CONNECTING"] = 0] = "CONNECTING";
    /** Connection is established and communication is possible */
    WebSocketReadyState[WebSocketReadyState["OPEN"] = 1] = "OPEN";
    /** Connection is going through the closing handshake */
    WebSocketReadyState[WebSocketReadyState["CLOSING"] = 2] = "CLOSING";
    /** Connection has been closed or could not be opened */
    WebSocketReadyState[WebSocketReadyState["CLOSED"] = 3] = "CLOSED";
})(WebSocketReadyState || (WebSocketReadyState = {}));
/**
 * Enum for standard relay error types
 */
export var RelayErrorType;
(function (RelayErrorType) {
    /** Connection error when attempting to connect to relay */
    RelayErrorType["CONNECTION_ERROR"] = "connection_error";
    /** Timeout occurred during a relay operation */
    RelayErrorType["TIMEOUT"] = "timeout";
    /** Connection was lost or closed during an operation */
    RelayErrorType["DISCONNECTED"] = "disconnected";
    /** Failed to validate an event (signature, id, etc.) */
    RelayErrorType["VALIDATION_ERROR"] = "validation_error";
    /** Event was rejected by the relay */
    RelayErrorType["EVENT_REJECTED"] = "event_rejected";
    /** Rate limit exceeded on the relay */
    RelayErrorType["RATE_LIMITED"] = "rate_limited";
    /** Authentication required but not provided */
    RelayErrorType["AUTH_REQUIRED"] = "auth_required";
    /** Authentication attempt failed */
    RelayErrorType["AUTH_FAILED"] = "auth_failed";
    /** IP address is blocked or restricted */
    RelayErrorType["IP_BLOCKED"] = "ip_blocked";
    /** Relay is in read-only mode */
    RelayErrorType["READ_ONLY"] = "read_only";
    /** Event payload is too large */
    RelayErrorType["PAYLOAD_TOO_LARGE"] = "payload_too_large";
    /** Relay could not process the message */
    RelayErrorType["INVALID_MESSAGE"] = "invalid_message";
    /** Maximum subscriptions reached */
    RelayErrorType["MAX_SUBSCRIPTIONS"] = "max_subscriptions";
    /** Unknown error */
    RelayErrorType["UNKNOWN"] = "unknown";
})(RelayErrorType || (RelayErrorType = {}));
export var NIP20Prefix;
(function (NIP20Prefix) {
    NIP20Prefix["Duplicate"] = "duplicate:";
    NIP20Prefix["PoW"] = "pow:";
    NIP20Prefix["RateLimited"] = "rate-limited:";
    NIP20Prefix["Invalid"] = "invalid:";
    NIP20Prefix["Error"] = "error:";
    NIP20Prefix["Blocked"] = "blocked:";
    NIP20Prefix["AuthRequired"] = "auth-required:";
    NIP20Prefix["Restricted"] = "restricted:";
})(NIP20Prefix || (NIP20Prefix = {}));
