// snstr/dist/src/nip47/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionType = exports.ERROR_RECOVERY_HINTS = exports.ERROR_CATEGORIES = exports.NIP47ErrorCode = exports.NIP47ErrorCategory = exports.NIP47NotificationType = exports.NIP47Method = exports.NIP47EncryptionScheme = exports.NIP47EventKind = void 0;
// Event kinds defined in NIP-47
var NIP47EventKind;
(function (NIP47EventKind) {
    NIP47EventKind[NIP47EventKind["INFO"] = 13194] = "INFO";
    NIP47EventKind[NIP47EventKind["REQUEST"] = 23194] = "REQUEST";
    NIP47EventKind[NIP47EventKind["RESPONSE"] = 23195] = "RESPONSE";
    NIP47EventKind[NIP47EventKind["NOTIFICATION"] = 23196] = "NOTIFICATION";
    NIP47EventKind[NIP47EventKind["NOTIFICATION_NIP44"] = 23197] = "NOTIFICATION_NIP44";
})(NIP47EventKind || (exports.NIP47EventKind = NIP47EventKind = {}));
// Encryption schemes supported by NIP-47
var NIP47EncryptionScheme;
(function (NIP47EncryptionScheme) {
    NIP47EncryptionScheme["NIP04"] = "nip04";
    NIP47EncryptionScheme["NIP44_V2"] = "nip44_v2";
})(NIP47EncryptionScheme || (exports.NIP47EncryptionScheme = NIP47EncryptionScheme = {}));
// Methods defined in NIP-47
var NIP47Method;
(function (NIP47Method) {
    NIP47Method["GET_INFO"] = "get_info";
    NIP47Method["GET_BALANCE"] = "get_balance";
    NIP47Method["PAY_INVOICE"] = "pay_invoice";
    NIP47Method["MAKE_INVOICE"] = "make_invoice";
    NIP47Method["LOOKUP_INVOICE"] = "lookup_invoice";
    NIP47Method["LIST_TRANSACTIONS"] = "list_transactions";
    NIP47Method["SIGN_MESSAGE"] = "sign_message";
    // Extended methods (not in the NIP-47 standard)
    NIP47Method["PAY_KEYSEND"] = "pay_keysend";
    NIP47Method["MULTI_PAY_INVOICE"] = "multi_pay_invoice";
    NIP47Method["MULTI_PAY_KEYSEND"] = "multi_pay_keysend";
    // Used for early-reject errors where actual method is unknown
    NIP47Method["UNKNOWN"] = "unknown";
})(NIP47Method || (exports.NIP47Method = NIP47Method = {}));
// NIP-47 notification types
var NIP47NotificationType;
(function (NIP47NotificationType) {
    NIP47NotificationType["PAYMENT_RECEIVED"] = "payment_received";
    NIP47NotificationType["PAYMENT_SENT"] = "payment_sent";
})(NIP47NotificationType || (exports.NIP47NotificationType = NIP47NotificationType = {}));
// NIP-47 error categories for better organization
var NIP47ErrorCategory;
(function (NIP47ErrorCategory) {
    NIP47ErrorCategory["AUTHORIZATION"] = "AUTHORIZATION";
    NIP47ErrorCategory["VALIDATION"] = "VALIDATION";
    NIP47ErrorCategory["RESOURCE"] = "RESOURCE";
    NIP47ErrorCategory["NETWORK"] = "NETWORK";
    NIP47ErrorCategory["WALLET"] = "WALLET";
    NIP47ErrorCategory["TIMEOUT"] = "TIMEOUT";
    NIP47ErrorCategory["INTERNAL"] = "INTERNAL";
})(NIP47ErrorCategory || (exports.NIP47ErrorCategory = NIP47ErrorCategory = {}));
// NIP-47 error codes
var NIP47ErrorCode;
(function (NIP47ErrorCode) {
    // Standard NIP-47 error codes (from spec)
    NIP47ErrorCode["UNAUTHORIZED"] = "UNAUTHORIZED";
    NIP47ErrorCode["INVALID_REQUEST"] = "INVALID_REQUEST";
    NIP47ErrorCode["INSUFFICIENT_BALANCE"] = "INSUFFICIENT_BALANCE";
    NIP47ErrorCode["PAYMENT_FAILED"] = "PAYMENT_FAILED";
    NIP47ErrorCode["INVOICE_EXPIRED"] = "INVOICE_EXPIRED";
    NIP47ErrorCode["NOT_FOUND"] = "NOT_FOUND";
    NIP47ErrorCode["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    NIP47ErrorCode["UNSUPPORTED_ENCRYPTION"] = "UNSUPPORTED_ENCRYPTION";
    // Request expiration (mentioned in spec)
    NIP47ErrorCode["REQUEST_EXPIRED"] = "REQUEST_EXPIRED";
    // Extended error codes (implementation-specific)
    NIP47ErrorCode["TIMEOUT"] = "TIMEOUT";
    NIP47ErrorCode["CONNECTION_ERROR"] = "CONNECTION_ERROR";
    NIP47ErrorCode["RELAY_ERROR"] = "RELAY_ERROR";
    NIP47ErrorCode["PUBLISH_FAILED"] = "PUBLISH_FAILED";
    NIP47ErrorCode["ENCRYPTION_ERROR"] = "ENCRYPTION_ERROR";
    NIP47ErrorCode["DECRYPTION_ERROR"] = "DECRYPTION_ERROR";
    NIP47ErrorCode["UNAUTHORIZED_CLIENT"] = "UNAUTHORIZED_CLIENT";
    // More granular validation errors
    NIP47ErrorCode["INVALID_INVOICE_FORMAT"] = "INVALID_INVOICE_FORMAT";
    NIP47ErrorCode["INVALID_AMOUNT"] = "INVALID_AMOUNT";
    NIP47ErrorCode["INVALID_PARAMETER"] = "INVALID_PARAMETER";
    // More granular payment errors
    NIP47ErrorCode["PAYMENT_ROUTE_NOT_FOUND"] = "PAYMENT_ROUTE_NOT_FOUND";
    NIP47ErrorCode["PAYMENT_REJECTED"] = "PAYMENT_REJECTED";
    // More granular wallet errors
    NIP47ErrorCode["WALLET_LOCKED"] = "WALLET_LOCKED";
    NIP47ErrorCode["WALLET_UNAVAILABLE"] = "WALLET_UNAVAILABLE";
    // Method-specific error codes
    NIP47ErrorCode["LOOKUP_INVOICE_FAILED"] = "LOOKUP_INVOICE_FAILED";
    // Client lifecycle error codes
    NIP47ErrorCode["NOT_INITIALIZED"] = "NOT_INITIALIZED";
})(NIP47ErrorCode || (exports.NIP47ErrorCode = NIP47ErrorCode = {}));
// Error code to category mapping
exports.ERROR_CATEGORIES = {
    [NIP47ErrorCode.UNAUTHORIZED]: NIP47ErrorCategory.AUTHORIZATION,
    [NIP47ErrorCode.UNAUTHORIZED_CLIENT]: NIP47ErrorCategory.AUTHORIZATION,
    [NIP47ErrorCode.INVALID_REQUEST]: NIP47ErrorCategory.VALIDATION,
    [NIP47ErrorCode.INVALID_INVOICE_FORMAT]: NIP47ErrorCategory.VALIDATION,
    [NIP47ErrorCode.INVALID_AMOUNT]: NIP47ErrorCategory.VALIDATION,
    [NIP47ErrorCode.INVALID_PARAMETER]: NIP47ErrorCategory.VALIDATION,
    [NIP47ErrorCode.INSUFFICIENT_BALANCE]: NIP47ErrorCategory.RESOURCE,
    [NIP47ErrorCode.NOT_FOUND]: NIP47ErrorCategory.RESOURCE,
    [NIP47ErrorCode.WALLET_UNAVAILABLE]: NIP47ErrorCategory.RESOURCE,
    [NIP47ErrorCode.PAYMENT_FAILED]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.PAYMENT_ROUTE_NOT_FOUND]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.PAYMENT_REJECTED]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.INVOICE_EXPIRED]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.WALLET_LOCKED]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.LOOKUP_INVOICE_FAILED]: NIP47ErrorCategory.WALLET,
    [NIP47ErrorCode.TIMEOUT]: NIP47ErrorCategory.TIMEOUT,
    [NIP47ErrorCode.REQUEST_EXPIRED]: NIP47ErrorCategory.TIMEOUT,
    [NIP47ErrorCode.CONNECTION_ERROR]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.RELAY_ERROR]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.PUBLISH_FAILED]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.ENCRYPTION_ERROR]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.DECRYPTION_ERROR]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.UNSUPPORTED_ENCRYPTION]: NIP47ErrorCategory.NETWORK,
    [NIP47ErrorCode.INTERNAL_ERROR]: NIP47ErrorCategory.INTERNAL,
    [NIP47ErrorCode.NOT_INITIALIZED]: NIP47ErrorCategory.INTERNAL,
};
// Recovery hints for error codes
exports.ERROR_RECOVERY_HINTS = {
    [NIP47ErrorCode.UNAUTHORIZED]: "Check connection credentials and permissions",
    [NIP47ErrorCode.UNAUTHORIZED_CLIENT]: "This client is not authorized. Contact wallet administrator",
    [NIP47ErrorCode.INVALID_REQUEST]: "Check request parameters and try again",
    [NIP47ErrorCode.INSUFFICIENT_BALANCE]: "Add funds to your wallet and try again",
    [NIP47ErrorCode.PAYMENT_FAILED]: "Payment could not be processed. Try again later",
    [NIP47ErrorCode.INVOICE_EXPIRED]: "Request a new invoice and try again",
    [NIP47ErrorCode.NOT_FOUND]: "The requested resource was not found. For lookupInvoice, check that the payment_hash or invoice exists in the wallet database",
    [NIP47ErrorCode.INTERNAL_ERROR]: "An internal error occurred. Please report this issue",
    [NIP47ErrorCode.REQUEST_EXPIRED]: "Request timed out. Try again with a longer expiration time",
    [NIP47ErrorCode.TIMEOUT]: "Operation timed out. Check connection and try again",
    [NIP47ErrorCode.CONNECTION_ERROR]: "Could not connect to the relay. Check network and relay status",
    [NIP47ErrorCode.RELAY_ERROR]: "Error communicating with relay. Try a different relay",
    [NIP47ErrorCode.PUBLISH_FAILED]: "Failed to publish event to relay. Try again or use a different relay",
    [NIP47ErrorCode.ENCRYPTION_ERROR]: "Encryption error. Check encryption parameters",
    [NIP47ErrorCode.DECRYPTION_ERROR]: "Decryption error. Check if using correct keys",
    [NIP47ErrorCode.UNSUPPORTED_ENCRYPTION]: "Encryption scheme not supported by wallet service. Check supported schemes in info event",
    [NIP47ErrorCode.INVALID_INVOICE_FORMAT]: "Invoice format is invalid. Check the invoice",
    [NIP47ErrorCode.INVALID_AMOUNT]: "Amount is invalid. Must be a positive integer",
    [NIP47ErrorCode.INVALID_PARAMETER]: "One or more parameters are invalid",
    [NIP47ErrorCode.PAYMENT_ROUTE_NOT_FOUND]: "No route found to destination. Try a different recipient or amount",
    [NIP47ErrorCode.PAYMENT_REJECTED]: "Payment was rejected by recipient node",
    [NIP47ErrorCode.WALLET_LOCKED]: "Wallet is locked. Unlock wallet and try again",
    [NIP47ErrorCode.WALLET_UNAVAILABLE]: "Wallet is currently unavailable. Try again later",
    [NIP47ErrorCode.LOOKUP_INVOICE_FAILED]: "Failed to lookup invoice. Check payment hash or invoice and try again",
    [NIP47ErrorCode.NOT_INITIALIZED]: "Client not initialized. Call init() before using this method",
};
// Transaction types
var TransactionType;
(function (TransactionType) {
    TransactionType["INCOMING"] = "incoming";
    TransactionType["OUTGOING"] = "outgoing";
})(TransactionType || (exports.TransactionType = TransactionType = {}));
