// snstr/dist/src/nip57/client.js
"use strict";
/**
 * NIP-57 Zap Client implementation
 * Provides high-level functionality for working with zaps in Nostr
 *
 * The NostrZapClient provides a comprehensive interface for:
 * - Sending zaps to users and events
 * - Fetching and validating zap receipts
 * - Calculating zap statistics
 * - Working with zap splits
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZapClient = exports.NostrZapClient = void 0;
const event_1 = require("../nip01/event");
const time_1 = require("../utils/time");
const index_1 = require("./index");
const utils_1 = require("./utils");
const security_validator_1 = require("../utils/security-validator");
/**
 * Client for working with Lightning Zaps (NIP-57)
 *
 * This client provides high-level methods for sending zaps,
 * fetching zap receipts, and calculating zap statistics.
 */
class NostrZapClient {
    /**
     * Create a new zap client
     * @param options Options for configuring the client
     */
    constructor(options) {
        this.client = options.client;
        this.defaultRelays = options.defaultRelays || [];
    }
    /**
     * Private method to safely unsubscribe from a subscription
     *
     * @param subscriptionIds Array of subscription IDs
     * @param context Context string for error logging
     */
    cleanupSubscription(subscriptionIds, context = "cleanup") {
        try {
            if ((0, security_validator_1.validateArrayAccess)(subscriptionIds, 0)) {
                const subId = (0, security_validator_1.safeArrayAccess)(subscriptionIds, 0);
                if (typeof subId === "string") {
                    this.client.unsubscribe([subId]);
                }
            }
        }
        catch (error) {
            if (error instanceof security_validator_1.SecurityValidationError) {
                console.warn(`NIP-57: Bounds checking error in ${context}: ${error.message}`);
            }
        }
    }
    /**
     * Check if a user can receive zaps
     *
     * This method checks if a user's LNURL endpoint supports NIP-57 Nostr zaps.
     *
     * @param pubkey The user's public key
     * @param lnurl Optional LNURL from profile (if known)
     * @returns Whether the user can receive zaps
     */
    async canReceiveZaps(pubkey, lnurl) {
        const zapClient = new ZapClient({
            nostrClient: this.client,
            defaultRelays: this.defaultRelays,
        });
        return zapClient.canReceiveZaps(pubkey, lnurl);
    }
    /**
     * Send a zap to a user or event
     *
     * This method:
     * 1. Creates a zap request
     * 2. Sends it to the recipient's LNURL server
     * 3. Returns the invoice for payment
     *
     * Note: After getting the invoice, you need to pay it with a Lightning wallet.
     * The LNURL server will create and publish a zap receipt after payment.
     *
     * @param options Zap options including recipient, amount, and comment
     * @param privateKey Private key for signing the request
     * @returns Object with success status and invoice or error
     */
    async sendZap(options, privateKey) {
        const zapClient = new ZapClient({
            nostrClient: this.client,
            defaultRelays: this.defaultRelays,
        });
        // Get the invoice
        const result = await zapClient.getZapInvoice(options, privateKey);
        if (result.error) {
            return {
                success: false,
                error: result.error,
            };
        }
        return {
            success: true,
            invoice: result.invoice,
        };
        // Note: In a real implementation, you would now:
        // 1. Use a Lightning wallet to pay the invoice
        // 2. Return the payment success/failure
    }
    /**
     * Fetch zaps received by a user
     *
     * @param pubkey User's public key
     * @param options Filter options
     * @returns Array of zap receipt events
     */
    async fetchUserReceivedZaps(pubkey, options = {}) {
        const filter = {
            kinds: [9735],
            "#p": [pubkey],
            limit: options.limit || 20,
        };
        if (options.since) {
            filter.since = options.since;
        }
        if (options.until) {
            filter.until = options.until;
        }
        if (options.authors && options.authors.length > 0) {
            filter.authors = options.authors;
        }
        // Create a promise to collect events
        return new Promise((resolve) => {
            const events = [];
            // Subscribe to zap receipts
            const subscriptionIds = this.client.subscribe([filter], (event) => {
                events.push(event);
            }, () => {
                // On EOSE, resolve with collected events
                this.cleanupSubscription(subscriptionIds);
                resolve(events);
            });
            // Set a timeout in case EOSE never comes
            setTimeout(() => {
                this.cleanupSubscription(subscriptionIds, "timeout");
                resolve(events);
            }, 10000);
        });
    }
    /**
     * Fetch zaps sent by a user
     *
     * @param pubkey User's public key
     * @param options Filter options
     * @returns Array of zap receipt events
     */
    async fetchUserSentZaps(pubkey, options = {}) {
        // We need to find zap receipts where the zap request was from this pubkey
        // This is more complex and less reliable since we need to parse each description
        // First get all zap receipts within our criteria
        const allZapReceipts = await this.fetchZapReceipts(options);
        // Then filter to only those sent by this user
        return allZapReceipts.filter((zapReceipt) => {
            try {
                // Find description tag with streamlined validation
                const descriptionTag = zapReceipt.tags.find((tag) => Array.isArray(tag) && tag.length > 0 && tag[0] === "description");
                if (!Array.isArray(descriptionTag) ||
                    descriptionTag.length < 2 ||
                    typeof descriptionTag[1] !== "string")
                    return false;
                // Parse zap request
                const zapRequest = JSON.parse(descriptionTag[1]);
                // Check if the sender is the specified pubkey
                return zapRequest.pubkey === pubkey;
            }
            catch (e) {
                if (e instanceof security_validator_1.SecurityValidationError) {
                    console.warn(`NIP-57: Bounds checking error in zap filtering: ${e.message}`);
                }
                return false;
            }
        });
    }
    /**
     * Fetch zaps for a specific event
     *
     * @param eventId The event ID to get zaps for
     * @param options Filter options
     * @returns Array of zap receipt events
     */
    async fetchEventZaps(eventId, options = {}) {
        const filter = {
            kinds: [9735],
            "#e": [eventId],
            limit: options.limit || 20,
        };
        if (options.since) {
            filter.since = options.since;
        }
        if (options.until) {
            filter.until = options.until;
        }
        if (options.authors && options.authors.length > 0) {
            filter.authors = options.authors;
        }
        // Create a promise to collect events
        return new Promise((resolve) => {
            const events = [];
            // Subscribe to zap receipts
            const subscriptionIds = this.client.subscribe([filter], (event) => {
                events.push(event);
            }, () => {
                // On EOSE, resolve with collected events
                this.cleanupSubscription(subscriptionIds);
                resolve(events);
            });
            // Set a timeout in case EOSE never comes
            setTimeout(() => {
                this.cleanupSubscription(subscriptionIds, "timeout");
                resolve(events);
            }, 10000);
        });
    }
    /**
     * Fetch all zap receipts matching the filter criteria
     *
     * @param options Filter options
     * @returns Array of zap receipt events
     */
    async fetchZapReceipts(options = {}) {
        const filter = {
            kinds: [9735],
            limit: options.limit || 20,
        };
        if (options.since) {
            filter.since = options.since;
        }
        if (options.until) {
            filter.until = options.until;
        }
        if (options.authors && options.authors.length > 0) {
            filter.authors = options.authors;
        }
        if (options.events && options.events.length > 0) {
            filter["#e"] = options.events;
        }
        // Create a promise to collect events
        return new Promise((resolve) => {
            const events = [];
            // Subscribe to zap receipts
            const subscriptionIds = this.client.subscribe([filter], (event) => {
                events.push(event);
            }, () => {
                // On EOSE, resolve with collected events
                this.cleanupSubscription(subscriptionIds);
                resolve(events);
            });
            // Set a timeout in case EOSE never comes
            setTimeout(() => {
                this.cleanupSubscription(subscriptionIds, "timeout");
                resolve(events);
            }, 10000);
        });
    }
    /**
     * Validate a zap receipt
     *
     * This validates that:
     * 1. The receipt is properly formatted
     * 2. The signature is valid
     * 3. The description hash in the bolt11 invoice matches the zap request
     * 4. The amount matches between invoice and zap request
     *
     * @param zapReceipt The zap receipt event to validate
     * @param lnurlPubkey Public key of the LNURL server (if known)
     * @returns Validation result with details
     */
    validateZapReceipt(zapReceipt, lnurlPubkey) {
        return (0, index_1.validateZapReceipt)(zapReceipt, lnurlPubkey || zapReceipt.pubkey);
    }
    /**
     * Calculate total zaps received by a user
     *
     * @param pubkey User's public key
     * @param options Filter options
     * @returns Zap statistics
     */
    async getTotalZapsReceived(pubkey, options = {}) {
        const zaps = await this.fetchUserReceivedZaps(pubkey, options);
        if (zaps.length === 0) {
            return { total: 0, count: 0 };
        }
        // Initialize stats
        const stats = {
            total: 0,
            count: 0,
            largest: 0,
            smallest: Number.MAX_SAFE_INTEGER,
            average: 0,
            firstAt: Number.MAX_SAFE_INTEGER,
            latestAt: 0,
        };
        // Process each zap
        for (const zap of zaps) {
            const validation = this.validateZapReceipt(zap);
            if (validation.valid && validation.amount) {
                stats.total += validation.amount;
                stats.count++;
                // Update largest/smallest
                if (validation.amount > (stats.largest || 0)) {
                    stats.largest = validation.amount;
                }
                if (validation.amount < (stats.smallest || Number.MAX_SAFE_INTEGER)) {
                    stats.smallest = validation.amount;
                }
                // Update timestamps
                if (zap.created_at < stats.firstAt) {
                    stats.firstAt = zap.created_at;
                }
                if (zap.created_at > stats.latestAt) {
                    stats.latestAt = zap.created_at;
                }
            }
        }
        // Calculate average
        if (stats.count > 0) {
            stats.average = Math.floor(stats.total / stats.count);
        }
        // Reset smallest if no valid zaps were found
        if (stats.smallest === Number.MAX_SAFE_INTEGER) {
            stats.smallest = undefined;
        }
        // Reset timestamps if no valid zaps were found
        if (stats.firstAt === Number.MAX_SAFE_INTEGER) {
            stats.firstAt = undefined;
        }
        return stats;
    }
    /**
     * Calculate total zaps for a specific event
     *
     * @param eventId Event ID to calculate zaps for
     * @param options Filter options
     * @returns Zap statistics
     */
    async getTotalZapsForEvent(eventId, options = {}) {
        const zaps = await this.fetchEventZaps(eventId, options);
        if (zaps.length === 0) {
            return { total: 0, count: 0 };
        }
        // Initialize stats
        const stats = {
            total: 0,
            count: 0,
            largest: 0,
            smallest: Number.MAX_SAFE_INTEGER,
            average: 0,
            firstAt: Number.MAX_SAFE_INTEGER,
            latestAt: 0,
        };
        // Process each zap
        for (const zap of zaps) {
            const validation = this.validateZapReceipt(zap);
            if (validation.valid && validation.amount) {
                stats.total += validation.amount;
                stats.count++;
                // Update largest/smallest
                if (validation.amount > (stats.largest || 0)) {
                    stats.largest = validation.amount;
                }
                if (validation.amount < (stats.smallest || Number.MAX_SAFE_INTEGER)) {
                    stats.smallest = validation.amount;
                }
                // Update timestamps
                if (zap.created_at < stats.firstAt) {
                    stats.firstAt = zap.created_at;
                }
                if (zap.created_at > stats.latestAt) {
                    stats.latestAt = zap.created_at;
                }
            }
        }
        // Calculate average
        if (stats.count > 0) {
            stats.average = Math.floor(stats.total / stats.count);
        }
        // Reset smallest if no valid zaps were found
        if (stats.smallest === Number.MAX_SAFE_INTEGER) {
            stats.smallest = undefined;
        }
        // Reset timestamps if no valid zaps were found
        if (stats.firstAt === Number.MAX_SAFE_INTEGER) {
            stats.firstAt = undefined;
        }
        return stats;
    }
    /**
     * Parse zap split information from an event
     *
     * @param event The event to parse zap split information from
     * @returns Array of zap split recipients with weights
     */
    parseZapSplit(event) {
        return (0, index_1.parseZapSplit)(event);
    }
    /**
     * Calculate zap split amounts
     *
     * @param totalAmount Total amount to split in millisats
     * @param splitInfo Array of zap split recipients with weights
     * @returns Array of recipients with calculated amounts
     */
    calculateZapSplitAmounts(totalAmount, splitInfo) {
        return (0, index_1.calculateZapSplitAmounts)(totalAmount, splitInfo);
    }
}
exports.NostrZapClient = NostrZapClient;
/**
 * Client for working with NIP-57 Zaps
 */
class ZapClient {
    /**
     * Create a new ZapClient
     * @param options Options for the client
     */
    constructor(options) {
        this.lnurlCache = new Map();
        this.nostrClient = options.nostrClient;
        this.defaultRelays = options.defaultRelays || [];
    }
    /**
     * Check if a user has a valid LNURL that supports zaps
     * @param pubkey User's public key
     * @param lnurlFromProfile LNURL from the user's profile (optional)
     * @returns Whether the user can receive zaps
     */
    async canReceiveZaps(pubkey, lnurlFromProfile) {
        try {
            // Check cache first
            const cached = this.lnurlCache.get(pubkey);
            if (cached) {
                return cached.supportsZaps;
            }
            // Use provided LNURL or fetch from profile
            const lnurl = lnurlFromProfile;
            if (!lnurl) {
                // In a real implementation, we would fetch the user's profile to get their LNURL
                // For now, return false as we don't have profile fetching implemented
                return false;
            }
            // Fetch and validate LNURL metadata
            const metadata = await (0, utils_1.fetchLnurlPayMetadata)(lnurl);
            if (!metadata) {
                this.lnurlCache.set(pubkey, { pubkey, lnurl, supportsZaps: false });
                return false;
            }
            // Check if it supports zaps
            const supportsZaps = (0, utils_1.supportsNostrZaps)(metadata);
            // Cache the result
            this.lnurlCache.set(pubkey, { pubkey, lnurl, supportsZaps });
            return supportsZaps;
        }
        catch (error) {
            console.error("Error checking zap support:", error);
            return false;
        }
    }
    /**
     * Generate a zap invoice for a user or event
     * @param options Options for the zap
     * @param privateKey Private key to sign the zap request
     * @returns Zap invoice result
     */
    async getZapInvoice(options, privateKey) {
        try {
            const senderPubkey = this.nostrClient.getPublicKey();
            if (!senderPubkey && !options.anonymousZap) {
                return {
                    invoice: "",
                    zapRequest: {},
                    error: "No public key available and not anonymous zap",
                };
            }
            // Fetch LNURL metadata
            const metadata = await (0, utils_1.fetchLnurlPayMetadata)(options.lnurl);
            if (!metadata) {
                return {
                    invoice: "",
                    zapRequest: {},
                    error: "Invalid LNURL or failed to fetch metadata",
                };
            }
            // Check if LNURL supports zaps
            if (!(0, utils_1.supportsNostrZaps)(metadata)) {
                return {
                    invoice: "",
                    zapRequest: {},
                    error: "LNURL does not support Nostr zaps",
                };
            }
            // Check amount limits
            if (options.amount < metadata.minSendable ||
                options.amount > metadata.maxSendable) {
                return {
                    invoice: "",
                    zapRequest: {},
                    error: `Amount out of range (${metadata.minSendable}-${metadata.maxSendable} millisats)`,
                };
            }
            // Create zap request
            const zapRequestOptions = {
                recipientPubkey: options.recipientPubkey,
                amount: options.amount,
                relays: options.relays || this.defaultRelays,
                content: options.comment || "",
                lnurl: options.lnurl,
            };
            // Add optional parameters
            if (options.eventId) {
                zapRequestOptions.eventId = options.eventId;
            }
            if (options.aTag) {
                zapRequestOptions.aTag = options.aTag;
            }
            // For anonymous zaps
            if (options.anonymousZap && senderPubkey) {
                zapRequestOptions.senderPubkey = senderPubkey;
            }
            // Create and sign zap request
            const requestTemplate = (0, index_1.createZapRequest)(zapRequestOptions, options.anonymousZap
                ? "00000000000000000000000000000000000000000000000000000000000000"
                : senderPubkey || "");
            const signedZapRequest = await (0, event_1.createSignedEvent)({
                ...requestTemplate,
                tags: requestTemplate.tags || [],
                pubkey: options.anonymousZap
                    ? "00000000000000000000000000000000000000000000000000000000000000"
                    : senderPubkey || "",
                created_at: (0, time_1.getUnixTime)(),
            }, privateKey);
            // Build callback URL with the zap request
            const callbackUrl = (0, utils_1.buildZapCallbackUrl)(metadata.callback, JSON.stringify(signedZapRequest), options.amount);
            // Fetch invoice from LNURL
            const invoiceResponse = await fetch(callbackUrl);
            const invoiceData = (await invoiceResponse.json());
            if (invoiceData.status === "ERROR") {
                return {
                    invoice: "",
                    zapRequest: signedZapRequest,
                    error: invoiceData.reason || "LNURL error",
                };
            }
            return {
                invoice: invoiceData.pr,
                zapRequest: signedZapRequest,
                paymentHash: invoiceData.payment_hash,
                successAction: invoiceData.successAction,
            };
        }
        catch (error) {
            console.error("Error getting zap invoice:", error);
            return {
                invoice: "",
                zapRequest: {},
                error: `Error: ${error instanceof Error ? error.message : String(error)}`,
            };
        }
    }
    /**
     * Validate a zap receipt
     * @param zapReceipt Zap receipt event
     * @param lnurlPubkey LNURL server pubkey
     * @returns Validation result
     */
    validateZapReceipt(zapReceipt, lnurlPubkey) {
        return (0, index_1.validateZapReceipt)(zapReceipt, lnurlPubkey);
    }
    /**
     * Handle zap splits for an event
     * @param event Event with zap split tags
     * @param totalAmount Total amount to split
     * @returns Split information for each recipient
     */
    getZapSplitAmounts(event, totalAmount) {
        const splitInfo = (0, index_1.parseZapSplit)(event);
        return (0, index_1.calculateZapSplitAmounts)(totalAmount, splitInfo);
    }
}
exports.ZapClient = ZapClient;
